File: csgo-leagues-ios/csgo-leagues-ios/Data/Mappers/MatchMapper.swift
import Foundation

protocol MatchMappable {
    func toDomain(_ dto: MatchDTO) -> Match
}

struct MatchMapper {
    static func toDomain(_ dto: MatchDTO) -> Match {
        Match(
            id: dto.id,
            name: dto.name,
            status: mapStatus(dto.status),
            beginAt: parseDate(dto.beginAt),
            teams: dto.opponents.map { TeamMapper.toDomain($0.opponent) },
            league: LeagueMapper.toDomain(dto.league),
            serieName: dto.serie.fullName
        )
    }

    private static func mapStatus(_ raw: String) -> MatchStatus {
        switch raw {
        case "running":  return .inProgress
        case "finished": return .ended
        default:         return .scheduled
        }
    }

    private static let isoFormatterFull: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private static func parseDate(_ raw: String?) -> Date? {
        guard let raw else { return nil }
        if let date = isoFormatterFull.date(from: raw) { return date }
        return isoFormatter.date(from: raw)
    }
}

extension MatchMapper: MatchMappable {
    func toDomain(_ dto: MatchDTO) -> Match { MatchMapper.toDomain(dto) }
}

File: csgo-leagues-android/app/src/main/java/com/voidx/csgo/data/mapper/MatchMapper.kt
package com.voidx.csgo.data.mapper

import com.voidx.csgo.data.dto.MatchDTO
import com.voidx.csgo.domain.entity.Match
import com.voidx.csgo.domain.entity.MatchStatus
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

object MatchMapper {
    fun toDomain(dto: MatchDTO): Match =
        Match(
            id = dto.id,
            name = dto.name,
            status = dto.status.toDomainStatus(),
            beginAt = dto.beginAt.toLocalDateTime(),
            teams = dto.opponents.map { TeamMapper.toDomain(it.opponent) },
            league = LeagueMapper.toDomain(dto.league),
            serieName = dto.serie.fullName.orEmpty(),
        )

    private fun String.toDomainStatus(): MatchStatus = when (this) {
        "running" -> MatchStatus.InProgress
        "finished" -> MatchStatus.Ended
        else -> MatchStatus.Scheduled
    }

    private fun String?.toLocalDateTime(): LocalDateTime? {
        val raw = this?.takeIf(String::isNotBlank) ?: return null
        return runCatching {
            Instant.parse(raw)
                .atZone(ZoneId.systemDefault())
                .toLocalDateTime()
        }.getOrNull()
    }
}
